### ISSUE TITLE
Replace python-dbus with jeepney

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2768]
During #2744 we have: https://github.com/rockstor/rockstor-core/issues/2744#issuecomment-1810062592

> @FroggyFlox & @Hooverdan96 I've just found the following upstream comment: https://gitlab.freedesktop.org/dbus/dbus-python/-/issues/48#note_2045548 from two months ago, as of this comment.
> 
> > This project is only semi-maintained: merge requests welcome, but I am unlikely to be able to spend enough time on this to rewrite the tutorial.
> 
> As such we should look to limiting our use of this library, & ideally removing it in time.
> 
> I think users.py is currenlty the only user:
> 
> > lbuildvm:/opt/rockstor/src # grep -R "import dbus"
> > rockstor/system/users.py:import dbus

From @FroggyFlox comment https://github.com/rockstor/rockstor-core/issues/2728#issuecomment-1829931053

> Interesting that it pulls jeepney as a dependency. This actually is my current top candidate to replace python-dbus (https://github.com/rockstor/rockstor-core/issues/2744#issuecomment-1810062592).

So it looks like our existing secondary dependency on jeepney, from our direct dependency on keyring-pass, is a nice convenient replacement for python-dbus.

PyPi: https://pypi.org/project/jeepney/
0.8.0 Released: Apr 3, 2022
Requires: Python >=3.7

**Comment by FroggyFlox** [comment]
Thanks a lot for creating this issue. I thought I would share a few links from where my current interest in jeepney comes from:
https://gitlab.freedesktop.org/dbus/dbus-python/-/commit/f74338c72b79af840d4755a821e421688a9743fb
https://stackoverflow.com/a/76576713

The fact that `keyring` is using `jeepney` is reassuring and supports its use for our own needs. We'll have to establish some common functions to easily interact with it, but it seems to offer quite a few options: https://jeepney.readthedocs.io/en/latest/integrate.html

Of note, this issue should probably affect #2680 as we would thus rely on our `jeepney` use instead of `pystemd`.

**Comment by phillxnet** [comment]
@FroggyFlox Re:

> Of note, this issue should probably affect #2680 as we would thus rely on our `jeepney` use instead of `pystemd`.

Would we not use jeepney for DBUS interactions, and pystemd for systemd interactions?

**Comment by FroggyFlox** [comment]
> Would we not use jeepney for DBUS interactions, and pystemd for systemd interactions?

Maybe, yes. I need to get back into looking at these to refresh myself and give you an informed answer. I do remember something about pystemd related to its build dependency that using a pure-python solution like jeepney (needs verification) would alleviate. I remember building python-dbus adds a hard package dependency (or even build from source?) in our kiwi recipe if I'm correct. It would be nice to get away from that if possible.

Thanks for the reminder about this, though, as I did forget about it.

**Comment by FroggyFlox** [comment]
> Would we not use jeepney for DBUS interactions, and pystemd for systemd interactions?

This is becoming more and more enticing as working with jeepney is rather low-level and thus we may end up having to write our own wrappers around systemd to do what we need. Not impossible at all, but it seems using `pystemd` would basically import a library that does that for us already. Now the question of `pystemd`'s dependencies remains open, however, and this is where `jeepney` could have the advantage as it has no dependency, I believe. `pystemd` might require a system package as dependency, for reference... needs to investigate that.

A brief exploration of `jeepney`'s use to interrogate systemd unit, inspired heavily by: https://jeepney.readthedocs.io/en/stable/integrate.html

```python
destination = 'org.freedesktop.systemd1'
path = '/org/freedesktop/systemd1'
interface = 'org.freedesktop.systemd1.Manager'

from jeepney import DBusAddress, new_method_call
sysd = DBusAddress(path, destination, interface)

from jeepney.io.blocking import open_dbus_connection
connection = open_dbus_connection(bus="SYSTEM")
connection
message = new_method_call(sysd, 'ListUnits')
reply = connection.send_and_get_reply(message)
```

We can slightly simplify things by using Jeepney's autogeneration of wrappers using the DBus interfaces' introspection:

```shell
cd /opt/rockstor
poetry run python -m jeepney.bindgen --name org.freedesktop.systemd1 --path /org/freedesktop/systemd1
```

```python
from org_freedesktop_systemd1 import Manager
from jeepney.io.blocking import open_dbus_connection

# Instantiate the object
test = Manager()

# Connect to the SYSTEM bus and send the message, collecting the reply
connection = open_dbus_connection(bus='SYSTEM') # could also be 'SESSION'
reply = connection.send_and_get_reply(test.ListUnits())
reply.body[0]
```
The latter returns a long list of tuples such as:
```
(
'rockstor.service',
'Rockstor startup script',
'loaded',
'active',
'running',
'',
'/org/freedesktop/systemd1/unit/rockstor_2eservice',
0,
'',
'/'
)
```

**Comment by FroggyFlox** [comment]
Continuing writing down my attempts at getting familiar with the DBus interface and `jeepney` below...

To restart a service, we can use this main `Manager` object and the `RestartUnit` method. The [docs](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html) state:

> On reply, if successful, this method returns the newly created job object which has been enqueued for asynchronous activation. Callers that want to track the outcome of the actual start operation need to monitor the result of this job. This can be achieved in a race-free manner by first subscribing to the JobRemoved() signal, then calling StartUnit() and using the returned job object to filter out unrelated JobRemoved() signals, until the desired one is received, which will then carry the result of the start operation.

When running the following...
```python
reply = connection.send_and_get_reply(test.RestartUnit("smb.service", "replace"))
```
... we indeed get the `job` returned:
```python
>>> reply.body
('/org/freedesktop/systemd1/job/6711',)
```

We do see that the `smb` service is successfully restarted (it was already active and running):
```
Dec 31 10:44:22 buildvm systemd[1]: Stopping Samba SMB Daemon...
Dec 31 10:44:22 buildvm systemd[1]: smb.service: Deactivated successfully.
Dec 31 10:44:22 buildvm systemd[1]: Stopped Samba SMB Daemon.
Dec 31 10:44:22 buildvm systemd[1]: Starting Samba SMB Daemon...
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[520]: touch: cannot touch '/etc/apparmor.d/local/usr.sbin.smbd-shares': No such file or directory
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[521]: grep: /etc/apparmor.d/local/usr.sbin.smbd-shares: No such file or directory
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[478]: /usr/share/samba/update-apparmor-samba-profile: line 62: /etc/apparmor.d/local/usr.sbin.smbd-shares.new: No such file or directory
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[522]: diff: /etc/apparmor.d/local/usr.sbin.smbd-shares: No such file or directory
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[522]: diff: /etc/apparmor.d/local/usr.sbin.smbd-shares.new: No such file or directory
Dec 31 10:44:22 buildvm update-apparmor-samba-profile[523]: mv: cannot stat '/etc/apparmor.d/local/usr.sbin.smbd-shares.new': No such file or directory
Dec 31 10:44:22 buildvm smbd[526]: [2024/12/31 10:44:22.976029,  0] ../../source3/smbd/server.c:1748(main)
Dec 31 10:44:22 buildvm smbd[526]:   smbd version 4.19.8-git.368.51d32c069f150600.3.6.11SUSE-oS15.0-x86_64 started.
Dec 31 10:44:22 buildvm smbd[526]:   Copyright Andrew Tridgell and the Samba Team 1992-2023
Dec 31 10:44:22 buildvm systemd[1]: Started Samba SMB Daemon.
```

**Comment by FroggyFlox** [comment]
Reference for `systemd1` API, the D-Bus interface of systemd:
https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html



> The service manager exposes a number of objects on the bus: one `Manager` object as a central entry point for clients along with individual objects for each unit and for each queued job. The unit objects implement a generic Unit interface as well as a type-specific interface. For example, service units implement both `org.freedesktop.systemd1.Unit` and `org.freedesktop.system1.Service`. The manager object can list unit and job objects or directly convert a unit name or job identifier to a bus path of the corresponding D-Bus object.


Let's see if we find out how to interact with a specific unit more directly using the generic Unit interface.

Can `jeepney` autogenerate wrappers for us for this interface?
```
# cd /opt/rockstor
# poetry run python -m jeepney.bindgen --name org.freedesktop.systemd1 --path /org/freedesktop/systemd1/unit/smb_2eservice

Written 2 interface wrappers to org_freedesktop_systemd1_unit_smb_2eservice.py
```

The 2 wrappers that were generated here correspond to the `Service` and `Unit` interface for this path:
```python
class Service(MessageGenerator):
    interface = 'org.freedesktop.systemd1.Service'
    (...)
    
class Unit(MessageGenerator):
    interface = 'org.freedesktop.systemd1.Unit'
    (...)
```

These two have different methods and properties we can use, so let's look at their documentation:
 - `Service Unit` object: https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html#Service%20Unit%20Objects
 - `Unit` object: https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html#Unit%20Objects
 
 
> [!NOTE]
> The docs for the `Service Unit` objects state: "All service unit objects implement the `org.freedesktop.systemd1.Service` interface (described here) in addition to the generic `org.freedesktop.systemd1.Unit` interface (see above).". This is in addition to: "Most properties of the Service interface map directly to the corresponding settings in service unit files.".

We can also see that the methods for the `Unit` object allow operations that are already covered by the main `Manager` object (such as start/stop/restart/...). The docs mention:
https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html#Methods1

> Start(), Stop(), Reload(), Restart(), TryRestart(), ReloadOrRestart(), ReloadOrTryRestart(), Kill(), QueueSignal(), ResetFailed(), and SetProperties() implement the same operation as the respective methods on the Manager object (see above). However, these methods operate on the unit object and hence do not take a unit name parameter. Invoking the methods directly on the Manager object has the advantage of not requiring a GetUnit() call to get the unit object for a specific unit name. Calling the methods on the Manager object is hence a round trip optimization.



Let's see what we can do with this by following the same approach we used for the `Manager` object. As a goal here, let's try to get the "ACTIVE" state of the `smb` service. This would be the `ActiveState` property of the `Unit` object. Using `dbus-send`, this would achieved using:
```
buildvm:/opt/rockstor # dbus-send --system --print-reply --dest=org.freedesktop.systemd1 /org/freedesktop/systemd1/unit/smb_2eservice org.freedesktop.DBus.Properties.Get string:org.freedesktop.systemd1.Unit string:ActiveState
method return time=1735662834.105292 sender=:1.2 -> destination=:1.92 serial=4121 reply_serial=2
   variant       string "active"
```

Using `jeepney`, this would be:
```python
destination = 'org.freedesktop.systemd1'
path = '/org/freedesktop/systemd1/unit/smb_2eservice'
interface = 'org.freedesktop.DBus.Properties'

from jeepney import DBusAddress, new_method_call
sysd = DBusAddress(path, destination, interface)

from jeepney.io.blocking import open_dbus_connection
connection = open_dbus_connection(bus="SYSTEM")
message = new_method_call(sysd, 'Get', 'ss', body=('org.freedesktop.systemd1.Unit', 'ActiveState'))
reply = connection.send_and_get_reply(message)
```
We get:
```python
>>> reply.body
(('s', 'active'),)

# manually stop the service: `systemctl stop smb`
>>> reply = connection.send_and_get_reply(message)
>>> reply.body
(('s', 'inactive'),)
```

It seems to work!
Making `jeepney` wrappers seems a bit more complicated here (at least autogenerated ones), so we may need to write some manually, or just use the approach above.

**Comment by FroggyFlox** [comment]
> so we may need to write some manually

This can be done with something like the following:
Use the following class:
```python
class UnitProperties(MessageGenerator):
    interface = 'org.freedesktop.DBus.Properties'

    def __init__(self, service: str, object_path_base: str='/org/freedesktop/systemd1/unit',
                 bus_name: str='org.freedesktop.systemd1'):
        object_path = f"{object_path_base}/{service}_2eservice"
        super().__init__(object_path=object_path, bus_name=bus_name)

    def Get(self, iface_name: str, property_name: str):
        return new_method_call(self, 'Get', 'ss',
                               (f"{self.bus_name}.{iface_name}", property_name))
```

Then query:
```python
from org_freedesktop_systemd1_unit_ind_service import UnitProperties
from jeepney.io.blocking import open_dbus_connection

# Instantiate the object
test = UnitProperties(service="smb")

# Connect to the SYSTEM bus and send the message, collecting the reply
connection = open_dbus_connection(bus='SYSTEM') # could also be 'SESSION'
reply = connection.send_and_get_reply(test.Get(iface_name="Service", property_name="StatusText"))
```
```python
>>> reply.body[0]
('s', 'smbd: ready to serve connections...')
```

**Comment by phillxnet** [comment]
From the Gitlab [README](https://gitlab.freedesktop.org/dbus/dbus-python/-/blob/master/README) of dbus-python we have:

> Problems and alternatives
> =========================
> 
> dbus-python might not be the best D-Bus binding for you to use. dbus-python
> does not follow the principle of "In the face of ambiguity, refuse the
> temptation to guess", and can't be changed to not do so without seriously
> breaking compatibility.
> 
> In addition, it uses libdbus (which has known problems with multi-threaded
> use) and attempts to be main-loop-agnostic (which means you have to select
> a suitable main loop for your application).
> 
> Alternative ways to get your Python code onto D-Bus include, in no
> particular order:
> 
> * GDBus, part of the GIO module of `GLib`_, via GObject-Introspection and
>   `PyGI`_ (uses the GLib main loop and object model)
> 
> * `dasbus`_, a higher-level wrapper around GDBus
> 
> * QtDBus, part of `Qt`_, via `PyQt`_ (uses the Qt main loop and object model)
> 
> * `dbussy`_, a CTypes wrapper around libdbus using asyncio for event handling
> 
> * `dbus-next`_, a pure-Python implementation of the D-Bus protocol
>   using asyncio for event handling
> 
> * `dbus-fast`_, a fork of dbus-next emphasizing performance
> 
> * `jeepney`_, a pure-Python implementation of the D-Bus protocol with
>   support for several event loop models
> 
>
